proto/src/determined/api/v1/trial.proto

```diff
@@ -756,7 +766,16 @@ message CompareTrialsRequest {
   Scale scale = 7;
   // x-axis selection. Default is in batch.
   XAxis x_axis = 8;
+  // metric ids for the query
+  repeated string metric_ids = 9;
+  // The type of range for the query
+  RangeType range_type = 10;
+  // Integer range for the query
+  determined.common.v1.Int32FieldFilter integer_range = 11;
+  // Time range for the query
+  determined.common.v1.TimestampFieldFilter time_range = 12;
```

Comment:
I do agree that this has the advantage of being able to combine different filters together. However, I am not sure about this use case or implementation. Two separate thoughts:

1. It sounds like the direction we are trying to move in is to allow users to specify any custom metrics to view as an x or y axis point. I imagine a future iteration of this code being something like, instead of this bun `ColumnExpr` that I have :

`ColumnExpr("(metrics ->'validation_metrics' ->> 'epoch')::float8 as epoch").` that I have being some dynamic piece 


more so

```
for customMetric in xAxisLabels:
     q =  q.ColumnExpr("(metrics ->'validation_metrics' ->> 'customMetric')::float8 as customMetric").
   
```

so the issue I foresee with the implementation would be having to define a bunch of metrics in this proto file as separate filters. Which would not really be possible anyway since we would need to know what "custom" metrics users might want to define beforehand. 

2. I am not sure if we have a use case yet to show all metrics vs all possible x axis in the webui at this point now, so being able to define both a batch, time, and epoch range all at once might not necessarily gain us anything. The expectation is that users will be viewing and filtering on these fields completely separately within the UI view, never concurrently. 

Also ccing @hkang1 to ensure I am making sense. 

If either (1) or (2) is not true then I am happy to discuss. Although, I do feel like a separate field for each metric would become hard to maintain.

Reply:
@ioga, design has plans for going in the future where the user add custom charts, where they can specify chart type, and their own x-axis and y-axis. So it's possible for them to add a chart for `[T]accuracy vs epoch` and then later add another one for `[T]accuracy vs [T]loss`. It's not today, but wondering if we should intentionally be flexible here so we don't paint ourselves in the corner. Not sure if this is too crazy to do but something like this maybe?
```
xAxisRangeType - timestamp, epoch, record, batches, <who knows what else>
xAxisRangeStartNumber (do we want integer instead of number?)
xAxisRangeEndNumber
xAxisRangeStartTimestamp
xAxisRangeEndTimestamp

yAxisRangeType - same as xAxis
<yAxis same versions as xAxis>
```

Reply:
that's a good commentary, but the enum path does not support custom columns either.

Reply:
I think the approach in the PR still gives us the most flexibility for custom metrics. 

I imagine for now the bulk majority of custom metrics will be an `integer` (or `float`) so if we wanted to support fully custom metrics today we could have logic very similar to the PR as is that could look like:

```
If range_type == "time":
   // timestamp logic
else:
  // integer logic (or float logic)
```

In the future when we do want to support fully custom metrics, and we really want to be as flexible as possible I think ditching `range_type batch` and any "specific" range type would make the most sense. 

I imagine on the front end we could possibly try to intelligently determine the metric type or simply have the user select it and then I imagine calls to this API being 


``
// for some custom time metric
rangeType = range_type_time
timeRangeFilter = some filter values
```
```
// for some custom float metric
rangeType = range_type_float
floatRangeFilter = some filter values
```
```
// for some custom time metric
rangeType = range_type_integer
integerRangeFilter = some filter values
```

This is assuming we will only ever be filtering on one field, which I believe is the use case we want to support. And then this would truly give us the flexibility to filter on any custom metric that we need.

 The only addition we would need is adding, similar to what I have brought up, some `xAxisLabel` metric field to the API so users could specify what custom metric they want to query the data by and virtually everything else could stay as is in this PR. The field would allow us to decouple the "range_type" from the metric we are setting as the `xAxis`

Reply:
I see no way to specify the custom column name with the enums. 
- if "range_type" enum means a column (i.e. batches, end time, epoch, etc.), then the enum won't support custom, user-specified  x-axis called "apples".
- if "range_type" enum means data type (i.e. float/int/timestamp), then it's redundant because we could check which of (timeRangeFilter, floatRangeFilter, integerRangeFilter) is not nil instead.

how about adding a polymorphic filter primitive
1. reusable proto for whenever we need to filter by custom column
```
# common.proto
message PolymorphicFilter {
    string name = 1; // filtered field/column/metric name
    DoubleFieldFilter double = 2;
    Int32FieldFilter int32 = 3;
    TimestampFieldFilter time = 4;
}
```
2. some code in `bun_filters.go` to validate only one of (double, int32, time) is specified, and to map it to a query for simple model columns (e.g. batches). for custom metric columns, that'll have to be done in a custom way in this API call.

Pros:
- will support custom metrics
- reusable
- no custom parsing

Cons:
- since grpc-gateway doesn't allow repeated structs on GET endpoints, only one of these can be present on the request. this should be fine for `CompareTrials` because there's only one possible x-axis. if one is not enough, but we need a limited number, potentially we could also support it in an ugly way:
```
message CompareTrialsRequest {
  ...
  PolymorphicFilter filter1 = 21;
  PolymorphicFilter filter2 = 22;
  PolymorphicFilter filter3 = 23;
  ...
}
```

Reply:
We don't need to support custom metrics in this version, we just need to make it easy to add that when we do so.